File: src/query/line.js
import csv from "papaparse";
import { sow } from "../record.js";
import { unescapeNewline } from "../escape.js";

export function queryLine(tablet, grains, stateInitial, state, line) {
  if (line === "") return state;

  const {
    data: [[fstEscaped, sndEscaped]],
  } = csv.parse(line, { delimiter: "," });

  const fst = unescapeNewline(fstEscaped);

  const snd = unescapeNewline(sndEscaped);

  const fstIsNew = state.fst === undefined || state.fst !== fst;

  state.fst = fst;

  const pushEndOfGroup = fstIsNew && state.isMatch;

  // if previous group of keys ended
  // move current state to "last"
  // and reset to initial state
  if (pushEndOfGroup) {
    const stateToPush = {
      entry: state.entry,
      query: state.query,
      thingQuerying: state.thingQuerying,
    };

    state.last = stateToPush;

    state.entry = { ...stateInitial.entry };

    state.query = { ...stateInitial.query };

    state.isMatch = false;
  }

  const trait = tablet.traitIsFirst ? fst : snd;

  const thing = tablet.thingIsFirst ? fst : snd;

  // this represents the current line
  // and will be sown into entry
  // if any of query grains match it
  const grainNew = {
    _: tablet.base,
    [tablet.trait]: trait,
    [tablet.thing]: thing,
  };

  // all grains need to match
  const isMatchGrains = grains.reduce((withGrain, grain) => {
    const isMatchGrain = tablet.traitIsRegex
      ? new RegExp(grain[tablet.trait]).test(trait)
      : grain[tablet.trait] === trait;

    const bothMatch =
      withGrain === undefined ? isMatchGrain : withGrain && isMatchGrain;

    return bothMatch;
  }, undefined);

  // NOTE: what the hell is thingQuerying?
  // it's kind of like telling next level of nesting
  // about what is currently being searched for

  // when querying also match literal trait from the query
  // otherwise always true
  const doDiff = stateInitial.thingQuerying !== undefined;

  const isMatchQuerying = doDiff ? stateInitial.thingQuerying === thing : true;

  const isMatch = isMatchGrains && isMatchQuerying;

  // when does state's isMatch overlap with current?
  state.isMatch = state.isMatch ? state.isMatch : isMatch;

  if (state.isMatch) {
    state.thingQuerying = thing;
  }

  if (isMatch) {
    state.entry = sow(state.entry, grainNew, tablet.trait, tablet.thing);

    state.query = sow(state.query, grainNew, tablet.trait, tablet.thing);
  }

  return state;
}

File: src/dataset/option/line.rs
pub fn option_line(tablet: Tablet, state: State, line: Line) -> Result<DoubleBuffer> {
    let mut last = None;

    let mut state = state;

    let fst_is_new = state.fst.is_none() || state.fst.as_ref() != Some(&line.key);

    state.fst = line.key;

    let push_end_of_group = fst_is_new && state.is_match;

    if push_end_of_group {
        last = Some(state);

        state.entry = Entry { base: tablet.base };

        state.is_match = false;
    }

    let base = if tablet.thing_is_first {
        line.key.to_owned()
    } else {
        line.value.to_owned()
    };

    let grain_new = Grain {
        base: tablet.base.to_owned(),
        base_value: Some(base),
        leaf: tablet.base.to_owned(),
        leaf_value: None,
    };

    let re_str = base;

    let re = Regex::new(&re_str)?;

    let is_match = re.is_match(state.query.base_value);

    let match_is_new = match state.match_map {
        None => true,
        Some(m) => m.get(&thing).is_none(),
    };

    state.is_match = if state.is_match {
        state.is_match
    } else {
        is_match && match_is_new
    };

    if is_match && match_is_new {
        match state.match_map.as_mut() {
            None() => (),
            Some(m) => m.insert(base, true);
        };

        // if previous querying tablet already matched thing
        // the trait in this record is likely to be the same
        // and might duplicate in the entry after sow
        let is_new_thing = match &state_initial.thing_querying {
            None => true,
            Some(t) => t != &thing,
        };

        if is_new_thing {
            state.query = match &state.query {
                None => panic!("unreachable"),
                Some(q) => Some(q.sow(&grain_new, &tablet.trait_, &tablet.thing)),
            };
        }
    }

    Ok(DoubleBuffer {
        current: state,
        last
    })
}
